File: xcode/samples/basic/QUICClient/QUICClient/Runtime/Transport/ITransporter.swift
//
//  ITransporter.swift
//  QUICClient
//
//  Created by Junior Silva (EXT) on 17/01/25.
//

import Foundation

@objc public protocol ITransporter {
    func connectQUIC(host: String, port: UInt16, handshakeOpts: String, completionHandler: @escaping (String) -> Void)
    func sendQUIC(messageType: MessageType, route: String, sequenceNumber: UInt, data: Data?, requestUid: UInt, timeout: Int, completionHandler: @escaping (String) -> Void)
    func disconnect() -> String
}

@objc public class Transporter: NSObject, ITransporter {

    private let quicClient = FrameworkQUICClient.shared

    public func connectQUIC(host: String, port: UInt16, handshakeOpts: String, completionHandler: @escaping (String) -> Void) {
        quicClient.connectQUIC(host: host, port: port, handshakeOpts: handshakeOpts, completionHandler: completionHandler)
    }

    public func sendQUIC(messageType: MessageType, route: String, sequenceNumber: UInt, data: Data?, requestUid: UInt, timeout: Int, completionHandler: @escaping (String) -> Void) {
        quicClient.sendQUIC(messageType: messageType.rawValue, route: route, sequenceNumber: sequenceNumber, data: data, requestUid: requestUid, timeout: timeout, completionHandler: completionHandler)
    }

    public func disconnect() -> String {
        return quicClient.disconnect()
    }
}

File: Unity/Packages/com.wildlifestudios.pitaya-client-csharp/Runtime/Transport/ITransporter.cs
using System;
using Wildlife.PitayaCSharp.Protocol;
using Wildlife.PitayaCSharp.Client;
using Wildlife.PitayaCSharp.Serializer;
using Wildlife.PitayaCSharp.Event;

namespace Wildlife.PitayaCSharp.Transport
{
    public interface ITransporter
    {
        IPitayaClient Client { get; }
        IPitayaClientConfig ClientConfig { get; }
        int Quality { get; }
        ProtobufSerializer.SerializationFormat? Serializer { get; }
        ITransporterPlugin Plugin { get; }
        TransportState State { get; }

        void Connect(string host, int port, string handshakeOpts);
        void Send(MessageType messageType, string route, uint sequenceNumber, byte[] data, uint requestUid, int timeout);
        void Send(byte[] packetBuffer, uint sequenceNumber, uint requestUid, int timeout); // This shouldn't be exposed here, but in a deeper layer.
        void Disconnect();

        // Consider the need to separate concerns
        void AddEventHandler(Action<IClientEvent> onClientEvent);
        void SetPushHandler(Action<string, byte[]> onPush);
    }
}
